Cluster Broccoli crashing after startup

Looks like there is some race condition on creating a websocket connection using the same session which leads to attempting to create the same websocket connection twice?

It seems to happen if you open two tabs in the same browser, the second one doesn't get a successful connection so maybe we need a set of connections for each user and allow creating another one for the same user?

```
Caused by: java.lang.IllegalArgumentException: ID 1hf6u'4hn*e'on~)_fjz02ue*b)yxu(0yng834~tp4ulmhr26*54'qd2m3lr_c(s is already an open web socket connection
	at de.frosner.broccoli.services.WebSocketService.newConnection(WebSocketService.scala:69) ~[cluster-broccoli.cluster-broccoli-0.6.0-sans-externalized.jar:na]
	at de.frosner.broccoli.controllers.WebSocketController$$anonfun$requestToSocket$1.apply(WebSocketController.scala:35) ~[cluster-broccoli.cluster-broccoli-0.6.0-sans-externalized.jar:na]
	at de.frosner.broccoli.controllers.WebSocketController$$anonfun$requestToSocket$1.apply(WebSocketController.scala:33) ~[cluster-broccoli.cluster-broccoli-0.6.0-sans-externalized.jar:na]
	at jp.t2v.lab.play2.auth.BroccoliWebsocketSecurity$$anonfun$withSecurity$2.apply(BroccoliWebsocketSecurity.scala:39) ~[cluster-broccoli.cluster-broccoli-0.6.0-sans-externalized.jar:na]
	at jp.t2v.lab.play2.auth.BroccoliWebsocketSecurity$$anonfun$withSecurity$2.apply(BroccoliWebsocketSecurity.scala:35) ~[cluster-broccoli.cluster-broccoli-0.6.0-sans-externalized.jar:na]
	at scala.concurrent.Future$$anonfun$flatMap$1.apply(Future.scala:251) ~[org.scala-lang.scala-library-2.10.6.jar:na]
	at scala.concurrent.Future$$anonfun$flatMap$1.apply(Future.scala:249) ~[org.scala-lang.scala-library-2.10.6.jar:na]
	at scala.concurrent.impl.CallbackRunnable.run(Promise.scala:32) [org.scala-lang.scala-library-2.10.6.jar:na]
	at scala.concurrent.impl.ExecutionContextImpl$$anon$3.exec(ExecutionContextImpl.scala:107) ~[org.scala-lang.scala-library-2.10.6.jar:na]
	... 4 common frames omitted
```

--- a/app/de/frosner/broccoli/controllers/SecurityController.scala
+++ b/app/de/frosner/broccoli/controllers/SecurityController.scala
@@ -32,7 +32,7 @@ case class SecurityController @Inject()
   }
 
   def login = Action.async { implicit request =>
-    getSessionId(request).map(id => (id, webSocketService.closeConnection(id))) match {
+    getSessionId(request).map(id => (id, webSocketService.closeConnections(id))) match {
       case Some((id, true)) => Logger.info(s"Removing websocket connection of $id due to another login")
       case _ =>
     }
@@ -73,7 +73,7 @@ case class SecurityController @Inject()
 
   def logout = Action.async { implicit request =>
     gotoLogoutSucceeded.andThen { case tryResult =>
-      getSessionId(request).map(id => (id, webSocketService.closeConnection(id))) match {
+      getSessionId(request).map(id => (id, webSocketService.closeConnections(id))) match {
         case Some((id, true)) => Logger.info(s"Removing websocket connection of $id due to logout")
         case Some((id, false)) => Logger.info(s"There was no websocket connection for session $id")
         case None => Logger.info(s"No session available to logout from")

--- a/app/de/frosner/broccoli/controllers/WebSocketController.scala
+++ b/app/de/frosner/broccoli/controllers/WebSocketController.scala
@@ -70,7 +70,7 @@ case class WebSocketController @Inject()
             }
             webSocketService.send(connectionId, Json.toJson(result))
         }.map { _ =>
-          webSocketService.closeConnection(connectionId)
+          webSocketService.closeConnections(connectionId)
           Logger.info(s"Closed connection $connectionLogString")
         }
 

--- a/app/de/frosner/broccoli/services/WebSocketService.scala
+++ b/app/de/frosner/broccoli/services/WebSocketService.scala
@@ -54,7 +54,7 @@ class WebSocketService @Inject() (templateService: TemplateService,
   private val scheduledTask = scheduler.scheduleAtFixedRate(task, 0, 1, TimeUnit.SECONDS)
 
   @volatile
-  private var connections: Map[String, (Account, Concurrent.Channel[Msg])] = Map.empty
+  private var connections: Map[String, (Account, Set[Concurrent.Channel[Msg]])] = Map.empty
 
   def newConnection(user: Account): (String, Enumerator[Msg]) = {
     val id = UUID.randomUUID().toString
@@ -62,22 +62,22 @@ class WebSocketService @Inject() (templateService: TemplateService,
   }
 
   def newConnection(id: String, user: Account): Enumerator[Msg] = {
-    val (enumerator, channel) = Concurrent.broadcast[Msg]
     if (connections.contains(id)) {
-      val error = s"ID $id is already an open web socket connection"
-      Logger.warn(error)
-      throw new IllegalArgumentException(error)
-    } else {
-      val (enumerator, channel) = Concurrent.broadcast[Msg]
-      connections = connections.updated(id, (user, channel))
-      enumerator
+      Logger.info(s"ID $id (${user.name}) already has an open web socket connection. Probably he/she has two tabs open?")
     }
+    val (enumerator, channel) = Concurrent.broadcast[Msg]
+    val maybeExistingChannels = connections.get(id)
+    val existingChannels = maybeExistingChannels.map {
+      case (u, c) => c
+    }.getOrElse(Set.empty)
+    connections = connections.updated(id, (user, existingChannels + channel))
+    enumerator
   }
 
-  def closeConnection(id: String): Boolean = {
+  def closeConnections(id: String): Boolean = {
     connections.get(id).exists {
-      case (user, channel) =>
-        channel.eofAndEnd()
+      case (user, channels) =>
+        channels.foreach(_.eofAndEnd())
         connections = connections - id
         true
     }
@@ -85,11 +85,11 @@ class WebSocketService @Inject() (templateService: TemplateService,
 
   def broadcast(msg: Account => Msg): Unit = {
     connections.foreach {
-      case (id, (user, channel)) =>
+      case (id, (user, channels)) =>
         // TODO we're kinda swallowing the exceptions here. That's ok because currently we have no authorization on broadcasts
         val tryMsg = Try {
           val actualMsg = msg(user)
-          channel.push(actualMsg)
+          channels.foreach(_.push(actualMsg))
           actualMsg
         }
         tryMsg match {
@@ -101,7 +101,7 @@ class WebSocketService @Inject() (templateService: TemplateService,
 
   def send(id: String, msg: Msg): Unit = {
     connections.get(id) match {
-      case Some((user, channel)) => channel.push(msg)
+      case Some((user, channels)) => channels.foreach(_.push(msg))
       case None => throw InvalidWebsocketConnectionException(id, connections.keys)
     }
   }

--- a/test/de/frosner/broccoli/controllers/SecurityControllerSpec.scala
+++ b/test/de/frosner/broccoli/controllers/SecurityControllerSpec.scala
@@ -90,7 +90,7 @@ class SecurityControllerSpec extends PlaySpecification with AuthUtils {
       )
       val result = controller.login.apply(FakeRequest().withLoggedIn(controller)(account.name))
       Await.ready(result, Duration(5, TimeUnit.SECONDS))
-      verify(controller.webSocketService).closeConnection(anyString())
+      verify(controller.webSocketService).closeConnections(anyString())
     }
 
     "don't cancel anything if it is the first login" in new WithApplication {
@@ -101,7 +101,7 @@ class SecurityControllerSpec extends PlaySpecification with AuthUtils {
       val requestWithData = FakeRequest().withMultipartFormDataBody(loginFormData(account.name, account.password))
       val result = controller.login.apply(requestWithData)
       Await.ready(result, Duration(5, TimeUnit.SECONDS))
-      verify(controller.webSocketService, times(0)).closeConnection(anyString())
+      verify(controller.webSocketService, times(0)).closeConnections(anyString())
     }
 
   }
@@ -125,7 +125,7 @@ class SecurityControllerSpec extends PlaySpecification with AuthUtils {
       )
       val result = controller.logout.apply(FakeRequest().withLoggedIn(controller)(account.name))
       Await.ready(result, Duration(5, TimeUnit.SECONDS))
-      verify(controller.webSocketService).closeConnection(anyString())
+      verify(controller.webSocketService).closeConnections(anyString())
     }
 
   }

